// scejs/dist/scejs/Prefabs/SimpleNode/SimpleNode.class.js
(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

var exp = { "WebCLGL": "./WebCLGL.class",
            "WebCLGLBuffer": "./WebCLGLBuffer.class.js",
            "WebCLGLFor": "./WebCLGLFor.class.js",
            "WebCLGLKernel": "./WebCLGLKernel.class.js",
            "WebCLGLUtils": "./WebCLGLUtils.class.js",
            "WebCLGLVertexFragmentProgram": "./WebCLGLVertexFragmentProgram.class.js"
};

for(var key in exp)
    exports[key] = require(exp[key]);
},{}],2:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
* @class
* @constructor
*/
var Component = exports.Component = function () {
    function Component() {
        _classCallCheck(this, Component);

        this.type = null;
        this.node = null;
    }

    /**
     * initialize
     * @param {Node} nod
     * @param {WebGLRenderingContext} glCtx
     */


    _createClass(Component, [{
        key: "initialize",
        value: function initialize(nod, glCtx) {}

        /**
         * tick
         * @param {number} delta
         */

    }, {
        key: "tick",
        value: function tick(delta) {}
    }]);

    return Component;
}();

global.Component = Component;
module.exports.Component = Component;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],3:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.ComponentTransform = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Component2 = require("./Component.class");

var _Constants = require("./Constants");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
* @class
* @constructor
*/
var ComponentTransform = exports.ComponentTransform = function (_Component) {
	_inherits(ComponentTransform, _Component);

	function ComponentTransform() {
		_classCallCheck(this, ComponentTransform);

		var _this = _possibleConstructorReturn(this, (ComponentTransform.__proto__ || Object.getPrototypeOf(ComponentTransform)).call(this));

		_this.type = _Constants.Constants.COMPONENT_TYPES.TRANSFORM;
		_this.node = null;
		_this.gl = null;

		_this.mModelMatrix_Position = $M16([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
		_this.mModelMatrix_Rotation = $M16([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
		return _this;
	}

	/**
  * initialize
  * @param {Node} nod
  * @param {WebGLRenderingContext} glCtx.
  * @override
  */


	_createClass(ComponentTransform, [{
		key: "initialize",
		value: function initialize(nod, glCtx) {
			this.node = nod;
			this.gl = glCtx;
		}
	}, {
		key: "getMatrixPosition",


		/**
   * getMatrixPosition
   * @returns {StormM16}
   */
		value: function getMatrixPosition() {
			return this.mModelMatrix_Position;
		}
	}, {
		key: "getMatrixRotation",


		/**
   * getMatrixRotation
   * @returns {StormM16}
   */
		value: function getMatrixRotation() {
			return this.mModelMatrix_Rotation;
		}
	}]);

	return ComponentTransform;
}(_Component2.Component);

global.ComponentTransform = ComponentTransform;
module.exports.ComponentTransform = ComponentTransform;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"./Component.class":2,"./Constants":5}],4:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Component_GPU = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

require("webclgl");

var _Component2 = require("./Component.class");

var _Constants = require("./Constants");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @class
 * @constructor
 */
var Component_GPU = exports.Component_GPU = function (_Component) {
    _inherits(Component_GPU, _Component);

    function Component_GPU() {
        _classCallCheck(this, Component_GPU);

        var _this = _possibleConstructorReturn(this, (Component_GPU.__proto__ || Object.getPrototypeOf(Component_GPU)).call(this));

        _this.type = _Constants.Constants.COMPONENT_TYPES.GPU;
        _this.node = null;
        _this.gl = null;

        _this.gpufG = null;
        _this.args = {};
        return _this;
    }

    /**
     * initialize
     * @param {Node} nod
     * @param {WebGLRenderingContext} glCtx.
     * @override
     */


    _createClass(Component_GPU, [{
        key: "initialize",
        value: function initialize(nod, glCtx) {
            this.node = nod;
            this.gl = glCtx;
        }
    }, {
        key: "tick",


        /**
         * tick
         * @override
         */
        value: function tick(delta) {
            this.tickArguments();

            if (this.gpufG != null) this.gpufG.processKernels();

            if (this.gpufG != null) this.gpufG.processGraphic(undefined);
        }
    }, {
        key: "setGPUFor",


        /**
         * setGPUFor
         */
        value: function setGPUFor() {
            var _this2 = this,
                _arguments = arguments;

            for (var key in arguments[1]) {
                var expl = key.split(" ");
                if (expl != null && expl.length > 1) {
                    var argName = expl[1];
                    this.args[argName] = {
                        "fnvalue": arguments[1][key],
                        "updatable": null,
                        "splits": null,
                        "overrideDimensions": null };
                }
                arguments[1][key] = arguments[1][key]();
            }

            var F = function F() {
                return gpufor.apply(_this2, _arguments);
            };
            this.gpufG = new F();
        }
    }, {
        key: "getWebCLGL",


        /**
         * getWebCLGL
         * @returns {WebCLGL}
         */
        value: function getWebCLGL() {
            return this.gpufG.getWebCLGL();
        }
    }, {
        key: "addArgument",


        // █████╗ ██████╗  ██████╗ ██╗   ██╗███╗   ███╗███████╗███╗   ██╗████████╗███████╗
        //██╔══██╗██╔══██╗██╔════╝ ██║   ██║████╗ ████║██╔════╝████╗  ██║╚══██╔══╝██╔════╝
        //███████║██████╔╝██║  ███╗██║   ██║██╔████╔██║█████╗  ██╔██╗ ██║   ██║   ███████╗
        //██╔══██║██╔══██╗██║   ██║██║   ██║██║╚██╔╝██║██╔══╝  ██║╚██╗██║   ██║   ╚════██║
        //██║  ██║██║  ██║╚██████╔╝╚██████╔╝██║ ╚═╝ ██║███████╗██║ ╚████║   ██║   ███████║
        //╚═╝  ╚═╝╚═╝  ╚═╝ ╚═════╝  ╚═════╝ ╚═╝     ╚═╝╚══════╝╚═╝  ╚═══╝   ╚═╝   ╚══════╝

        /**
         *
         */
        value: function addArgument(arg, value) {
            this.args[arg.split(" ")[1]] = {
                "fnvalue": value,
                "updatable": null,
                "splits": null,
                "overrideDimensions": null };

            this.gpufG.addArg(arg);
        }
    }, {
        key: "setArg",


        /**
         * @param {String} argument Argument to set
         * @param {Function} fnvalue
         * @param {Array<number>} [splits=[array.length]]
         * @param {Array<number>} [overrideDimensions=new Array(){Math.sqrt(value.length), Math.sqrt(value.length)}]
         * @returns {WebCLGLBuffer}
         */
        value: function setArg(argument, fnvalue, splits, overrideDimensions) {
            var buff = this.gpufG.setArg(argument, fnvalue(), splits, overrideDimensions);
            this.args[argument] = { "fnvalue": fnvalue,
                "updatable": null,
                "splits": splits,
                "overrideDimensions": overrideDimensions };

            return buff;
        }
    }, {
        key: "getComponentBufferArg",


        /**
         * getComponentBufferArg
         * @param {String} argument Argument to set
         * @param {Component|Component_GPU} comp_gpu
         */
        value: function getComponentBufferArg(argument, comp_gpu) {
            this.gpufG.getGPUForArg(argument, comp_gpu.gpufG);
            this.args[argument] = { "fnvalue": null,
                "updatable": null,
                "splits": null,
                "overrideDimensions": null };
        }
    }, {
        key: "getArgs",


        /**
         * getArgs
         * @returns {Object}
         */
        value: function getArgs() {
            return this.args;
        }
    }, {
        key: "getAllArgs",


        /**
         * getAllArgs
         * @returns {Object}
         */
        value: function getAllArgs() {
            return this.gpufG.getAllArgs();
        }
    }, {
        key: "getBuffers",


        /**
         * getBuffers
         * @returns {{}|Array<WebCLGLBuffer>}
         */
        value: function getBuffers() {
            return this.gpufG._argsValues;
        }
    }, {
        key: "setArgUpdatable",


        /**
         * @param {String} argument Argument to set
         * @param {boolean} value
         */
        value: function setArgUpdatable(argument, value) {
            this.args[argument].updatable = value;
        }
    }, {
        key: "tickArguments",


        /**
         * tickArguments
         */
        value: function tickArguments() {
            for (var key in this.args) {
                if (this.args[key].updatable === true) {
                    var arg = this.args[key];
                    this.gpufG.setArg(key, arg.fnvalue(), arg.splits, arg.overrideDimensions);
                }
            }
        }
    }]);

    return Component_GPU;
}(_Component2.Component);

global.Component_GPU = Component_GPU;
module.exports.Component_GPU = Component_GPU;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"./Component.class":2,"./Constants":5,"webclgl":1}],5:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});
var Constants = exports.Constants = {
	"EVENT_TYPES": {
		"KEY_DOWN": 0,
		"KEY_UP": 1,
		"MOUSE_DOWN": 2,
		"MOUSE_UP": 3,
		"MOUSE_MOVE": 4,
		"MOUSE_WHEEL": 5
	},
	"COMPONENT_TYPES": {
		"KEYBOARD_EVENTS": 0,
		"MOUSE_EVENTS": 1,
		"TRANSFORM": 2,
		"TRANSFORM_TARGET": 3,
		"CONTROLLER_TRANSFORM_TARGET": 4,
		"PROJECTION": 5,
		"GPU": 6
	},
	"PROJECTION_TYPES": {
		"PERSPECTIVE": 0,
		"ORTHO": 1
	},
	"VIEW_TYPES": {
		"LEFT": 0,
		"RIGHT": 1,
		"FRONT": 2,
		"BACK": 3,
		"TOP": 4,
		"BOTTOM": 5
	},
	"BLENDING_MODES": {
		"ZERO": "ZERO",
		"ONE": "ONE",
		"SRC_COLOR": "SRC_COLOR",
		"ONE_MINUS_SRC_COLOR": "ONE_MINUS_SRC_COLOR",
		"DST_COLOR": "DST_COLOR",
		"ONE_MINUS_DST_COLOR": "ONE_MINUS_DST_COLOR",
		"SRC_ALPHA": "SRC_ALPHA",
		"ONE_MINUS_SRC_ALPHA": "ONE_MINUS_SRC_ALPHA",
		"DST_ALPHA": "DST_ALPHA",
		"ONE_MINUS_DST_ALPHA": "ONE_MINUS_DST_ALPHA",
		"SRC_ALPHA_SATURATE": "SRC_ALPHA_SATURATE",
		"CONSTANT_COLOR": "CONSTANT_COLOR",
		"ONE_MINUS_CONSTANT_COLOR": "ONE_MINUS_CONSTANT_COLOR",
		"CONSTANT_ALPHA": "CONSTANT_ALPHA",
		"ONE_MINUS_CONSTANT_ALPHA": "ONE_MINUS_CONSTANT_ALPHA"
	},
	"BLENDING_EQUATION_TYPES": {
		"FUNC_ADD": "FUNC_ADD",
		"FUNC_SUBTRACT": "FUNC_SUBTRACT",
		"FUNC_REVERSE_SUBTRACT": "FUNC_REVERSE_SUBTRACT"
	},
	"DRAW_MODES": {
		"POINTS": 0,
		"LINES": 1,
		"LINE_LOOP": 2,
		"LINE_STRIP": 3,
		"TRIANGLES": 4,
		"TRIANGLE_STRIP": 5,
		"TRIANGLE_FAN": 6
	}
};
global.Constants = Constants;
module.exports.Constants = Constants;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],6:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
        value: true
});
exports.SimpleNode = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _ComponentTransform = require("../../ComponentTransform.class");

var _Component_GPU = require("../../Component_GPU.class");

var _VFP_RGB = require("../../VFP_RGB.class");

var _Constants = require("../../Constants");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
* @class
*/
var SimpleNode = exports.SimpleNode = function () {
        function SimpleNode(sce) {
                _classCallCheck(this, SimpleNode);

                this._sce = sce;
                this._project = this._sce.getLoadedProject();

                this.node = new Node();
                this._project.getActiveStage().addNode(this.node);

                this._mesh = null;

                // ComponentTransform
                var comp_transform = new _ComponentTransform.ComponentTransform();
                this.node.addComponent(comp_transform);

                // Component_GPU
                this.comp_renderer = new _Component_GPU.Component_GPU();
                this.node.addComponent(this.comp_renderer);
        }

        /**
        * setMesh
        * @param {Mesh} mesh
        */


        _createClass(SimpleNode, [{
                key: "setMesh",
                value: function setMesh(mesh) {
                        var _this = this;

                        this._mesh = mesh;

                        this.comp_renderer.setGPUFor(this.comp_renderer.gl, { "float4*attr vertexPos": function float4AttrVertexPos() {
                                        return _this._mesh.vertexArray;
                                },
                                "float4*attr vertexNormal": function float4AttrVertexNormal() {
                                        return _this._mesh.normalArray;
                                },
                                "float4*attr vertexTexture": function float4AttrVertexTexture() {
                                        return _this._mesh.textureArray;
                                },
                                "float*attr vertexTextureUnit": function floatAttrVertexTextureUnit() {
                                        return _this._mesh.textureUnitArray;
                                },
                                "indices": function indices() {
                                        return _this._mesh.indexArray;
                                },
                                "mat4 PMatrix": function mat4PMatrix() {
                                        return _this._project.getActiveStage().getActiveCamera().getComponent(_Constants.Constants.COMPONENT_TYPES.PROJECTION).getMatrix().transpose().e;
                                },
                                "mat4 cameraWMatrix": function mat4CameraWMatrix() {
                                        return _this._project.getActiveStage().getActiveCamera().getComponent(_Constants.Constants.COMPONENT_TYPES.TRANSFORM_TARGET).getMatrix().transpose().e;
                                },
                                "mat4 nodeWMatrix": function mat4NodeWMatrix() {
                                        return _this.node.getComponent(_Constants.Constants.COMPONENT_TYPES.TRANSFORM).getMatrixPosition().transpose().e;
                                },
                                'float nodesSize': function floatNodesSize() {
                                        return 30.0;
                                },
                                'float4* texAlbedo': function float4TexAlbedo() {
                                        return _this._mesh.vertexArray;
                                }
                        }, { "type": "GRAPHIC",
                                "config": new _VFP_RGB.VFP_RGB(1).getSrc(),
                                "drawMode": 4,
                                "depthTest": true,
                                "blend": true,
                                "blendEquation": _Constants.Constants.BLENDING_EQUATION_TYPES.FUNC_ADD,
                                "blendSrcMode": _Constants.Constants.BLENDING_MODES.SRC_ALPHA,
                                "blendDstMode": _Constants.Constants.BLENDING_MODES.ONE_MINUS_SRC_ALPHA });
                        this.comp_renderer.setArgUpdatable("PMatrix", true);
                        this.comp_renderer.setArgUpdatable("cameraWMatrix", true);
                        this.comp_renderer.setArgUpdatable("nodeWMatrix", true);
                        this.comp_renderer.getComponentBufferArg("RGB", this._project.getActiveStage().getActiveCamera().getComponent(_Constants.Constants.COMPONENT_TYPES.GPU));
                        this.comp_renderer.gpufG.onPreProcessGraphic(0, function () {
                                //this.comp_renderer.gl.clear(this.comp_renderer.gl.COLOR_BUFFER_BIT | this.comp_renderer.gl.DEPTH_BUFFER_BIT);
                        });
                }
        }, {
                key: "setImage",


                /**
                * setImage
                * @param {String} url
                */
                value: function setImage(url) {
                        var _this2 = this;

                        var image = new Image();
                        image.onload = function () {
                                _this2.comp_renderer.setArg("texAlbedo", function () {
                                        return image;
                                });
                        };
                        image.src = url;
                }
        }]);

        return SimpleNode;
}();

global.SimpleNode = SimpleNode;
module.exports.SimpleNode = SimpleNode;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"../../ComponentTransform.class":3,"../../Component_GPU.class":4,"../../Constants":5,"../../VFP_RGB.class":7}],7:[function(require,module,exports){
(function (global){
'use strict';

Object.defineProperty(exports, "__esModule", {
									value: true
});
exports.VFP_RGB = VFP_RGB;
function VFP_RGB(textureUnitCount) {
									this.getSrc = function () {
																		return [["RGB"],
																		// vertex head
																		'varying vec4 vVN;\n' + 'varying vec4 vVT;\n' + 'varying float vVTU;\n',

																		// vertex source
																		'vec4 vp = vertexPos[];\n' + 'vec4 vn = vertexNormal[];\n' + 'vec4 vt = vertexTexture[];\n' + 'float vtu = vertexTextureUnit[];\n' + 'vVN = vn;' + 'vVT = vt;' + 'vVTU = vtu;' + 'gl_Position = PMatrix * cameraWMatrix * nodeWMatrix * vp;\n',

																		// fragment head
																		'varying vec4 vVN;\n' + 'varying vec4 vVT;\n' + 'varying float vVTU;\n',

																		// fragment source
																		'vec4 textureColor = texture2D(texAlbedo, vVT.xy);\n' +
																		// diffuse
																		'return [textureColor];\n'];
									};
}
global.VFP_RGB = VFP_RGB;
module.exports.VFP_RGB = VFP_RGB;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}]},{},[6]);
