// scejs/dist/scejs/ComponentTransformTarget.class.js
(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

var exp = { "WebCLGL": "./WebCLGL.class",
            "WebCLGLBuffer": "./WebCLGLBuffer.class.js",
            "WebCLGLFor": "./WebCLGLFor.class.js",
            "WebCLGLKernel": "./WebCLGLKernel.class.js",
            "WebCLGLUtils": "./WebCLGLUtils.class.js",
            "WebCLGLVertexFragmentProgram": "./WebCLGLVertexFragmentProgram.class.js"
};

for(var key in exp)
    exports[key] = require(exp[key]);
},{}],2:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
* @class
* @constructor
*/
var Component = exports.Component = function () {
    function Component() {
        _classCallCheck(this, Component);

        this.type = null;
        this.node = null;
    }

    /**
     * initialize
     * @param {Node} nod
     * @param {WebGLRenderingContext} glCtx
     */


    _createClass(Component, [{
        key: "initialize",
        value: function initialize(nod, glCtx) {}

        /**
         * tick
         * @param {number} delta
         */

    }, {
        key: "tick",
        value: function tick(delta) {}
    }]);

    return Component;
}();

global.Component = Component;
module.exports.Component = Component;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],3:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
							value: true
});
exports.ComponentTransformTarget = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Component2 = require("./Component.class");

var _Utils = require("./Utils.class");

var _StormMath = require("./StormMath.class");

var _Constants = require("./Constants");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
* @class
* @constructor
*/
var ComponentTransformTarget = exports.ComponentTransformTarget = function (_Component) {
							_inherits(ComponentTransformTarget, _Component);

							function ComponentTransformTarget() {
														_classCallCheck(this, ComponentTransformTarget);

														var _this = _possibleConstructorReturn(this, (ComponentTransformTarget.__proto__ || Object.getPrototypeOf(ComponentTransformTarget)).call(this));

														_this.type = _Constants.Constants.COMPONENT_TYPES.TRANSFORM_TARGET;
														_this.node = null;
														_this.gl = null;

														_this.mModelMatrix = $M16([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);

														_this.targetDistance = 5.0;
														_this.positionGoal = $V3([0, 0, _this.targetDistance]);
														_this.positionTarget = $V3([0, 0, 0]);
														return _this;
							}

							/**
        * initialize
        * @param {Node} nod
        * @param {WebGLRenderingContext} glCtx.
        * @override
        */


							_createClass(ComponentTransformTarget, [{
														key: "initialize",
														value: function initialize(nod, glCtx) {
																					this.node = nod;
																					this.gl = glCtx;

																					this.performMatrix();
														}
							}, {
														key: "getMatrix",


														/**
               * getMatrix
               * @returns {StormM16}
               */
														value: function getMatrix() {
																					return this.mModelMatrix;
														}
							}, {
														key: "setTargetDistance",


														/**
               * setTargetDistance
               * @param {number} distance
               */
														value: function setTargetDistance(distance) {
																					this.targetDistance = distance;
														}
							}, {
														key: "getTargetDistance",


														/**
               * getTargetDistance
               * @returns {number}
               */
														value: function getTargetDistance() {
																					return this.targetDistance;
														}
							}, {
														key: "setPositionGoal",


														/** 
               * setPositionGoal
               * @param {StormV3} position
               */
														value: function setPositionGoal(position) {
																					this.positionGoal = position;
																					this.performMatrix();
														}
							}, {
														key: "getPositionGoal",


														/** 
               * getPositionGoal
               * @returns {StormV3}
               */
														value: function getPositionGoal() {
																					return this.positionGoal;
														}
							}, {
														key: "setPositionTarget",


														/**
               * setPositionTarget
               * @param {StormV3} position
               */
														value: function setPositionTarget(position) {
																					this.positionTarget = position;
																					this.performMatrix();
														}
							}, {
														key: "getPositionTarget",


														/**
               * getPositionTarget
               * @returns {StormV3}
               */
														value: function getPositionTarget() {
																					return this.positionTarget;
														}
							}, {
														key: "reset",
														value: function reset() {
																					this.positionGoal = $V3([0, 0, this.targetDistance]);
																					this.positionTarget = $V3([0, 0, 0]);
																					this.performMatrix();
														}
							}, {
														key: "yaw",


														/**
               * yaw
               * @param {number} angle
               */
														value: function yaw(angle) {
																					var dir = this.getPositionGoal().subtract(this.getPositionTarget());

																					var sphericalCoords = _Utils.Utils.cartesianToSpherical(dir.normalize());
																					var cartesianCoords = _Utils.Utils.sphericalToCartesian(sphericalCoords.radius, sphericalCoords.lat, sphericalCoords.lng + angle);

																					this.setPositionGoal(this.getPositionTarget().add(cartesianCoords.x(this.getTargetDistance())));
														}
							}, {
														key: "pitch",


														/**
               * pitch
               * @param {number} angle
               */
														value: function pitch(angle) {
																					var dir = this.getPositionGoal().subtract(this.getPositionTarget());

																					var sphericalCoords = _Utils.Utils.cartesianToSpherical(dir.normalize());
																					var cartesianCoords = _Utils.Utils.sphericalToCartesian(sphericalCoords.radius, sphericalCoords.lat + angle, sphericalCoords.lng);

																					this.setPositionGoal(this.getPositionTarget().add(cartesianCoords.x(this.getTargetDistance())));
														}
							}, {
														key: "performMatrix",


														/**
               * performMatrix
               * @private
               */
														value: function performMatrix() {
																					this.mModelMatrix = _StormMath.StormM16.makeLookAt(this.getPositionGoal().e[0], this.getPositionGoal().e[1], this.getPositionGoal().e[2], this.getPositionTarget().e[0], this.getPositionTarget().e[1], this.getPositionTarget().e[2], 0, 1, 0);
														}
							}]);

							return ComponentTransformTarget;
}(_Component2.Component);

global.ComponentTransformTarget = ComponentTransformTarget;
module.exports.ComponentTransformTarget = ComponentTransformTarget;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"./Component.class":2,"./Constants":4,"./StormMath.class":5,"./Utils.class":6}],4:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});
var Constants = exports.Constants = {
	"EVENT_TYPES": {
		"KEY_DOWN": 0,
		"KEY_UP": 1,
		"MOUSE_DOWN": 2,
		"MOUSE_UP": 3,
		"MOUSE_MOVE": 4,
		"MOUSE_WHEEL": 5
	},
	"COMPONENT_TYPES": {
		"KEYBOARD_EVENTS": 0,
		"MOUSE_EVENTS": 1,
		"TRANSFORM": 2,
		"TRANSFORM_TARGET": 3,
		"CONTROLLER_TRANSFORM_TARGET": 4,
		"PROJECTION": 5,
		"GPU": 6
	},
	"PROJECTION_TYPES": {
		"PERSPECTIVE": 0,
		"ORTHO": 1
	},
	"VIEW_TYPES": {
		"LEFT": 0,
		"RIGHT": 1,
		"FRONT": 2,
		"BACK": 3,
		"TOP": 4,
		"BOTTOM": 5
	},
	"BLENDING_MODES": {
		"ZERO": "ZERO",
		"ONE": "ONE",
		"SRC_COLOR": "SRC_COLOR",
		"ONE_MINUS_SRC_COLOR": "ONE_MINUS_SRC_COLOR",
		"DST_COLOR": "DST_COLOR",
		"ONE_MINUS_DST_COLOR": "ONE_MINUS_DST_COLOR",
		"SRC_ALPHA": "SRC_ALPHA",
		"ONE_MINUS_SRC_ALPHA": "ONE_MINUS_SRC_ALPHA",
		"DST_ALPHA": "DST_ALPHA",
		"ONE_MINUS_DST_ALPHA": "ONE_MINUS_DST_ALPHA",
		"SRC_ALPHA_SATURATE": "SRC_ALPHA_SATURATE",
		"CONSTANT_COLOR": "CONSTANT_COLOR",
		"ONE_MINUS_CONSTANT_COLOR": "ONE_MINUS_CONSTANT_COLOR",
		"CONSTANT_ALPHA": "CONSTANT_ALPHA",
		"ONE_MINUS_CONSTANT_ALPHA": "ONE_MINUS_CONSTANT_ALPHA"
	},
	"BLENDING_EQUATION_TYPES": {
		"FUNC_ADD": "FUNC_ADD",
		"FUNC_SUBTRACT": "FUNC_SUBTRACT",
		"FUNC_REVERSE_SUBTRACT": "FUNC_REVERSE_SUBTRACT"
	},
	"DRAW_MODES": {
		"POINTS": 0,
		"LINES": 1,
		"LINE_LOOP": 2,
		"LINE_STRIP": 3,
		"TRIANGLES": 4,
		"TRIANGLE_STRIP": 5,
		"TRIANGLE_FAN": 6
	}
};
global.Constants = Constants;
module.exports.Constants = Constants;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],5:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.$M16 = $M16;
exports.$V3 = $V3;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/*
StormM16.rotation
Copyright (C) 2007 James Coglan Sylvester.js

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
*/

/*
StormM16.inverse
Copyright (C) 2004 - Geotechnical Software Services
 
This code is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public 
License as published by the Free Software Foundation; either 
version 2.1 of the License, or (at your option) any later version.

This code is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public 
License along with this program; if not, write to the Free 
Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, 
MA  02111-1307, USA.
*/

/**
* @class
* @constructor

* @property {Float32Array} e Array with 16 values
*/
var StormM16 = exports.StormM16 = function () {
    function StormM16(elements) {
        _classCallCheck(this, StormM16);

        this.e = null;

        if (elements !== undefined && elements !== null) this.e = new Float32Array(elements);else this.e = new Float32Array([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
    }

    /**
    * Perform a * b and get the matrix
    * @returns {StormM16}
    * @param {StormM16|StormV3} stormM16
    */


    _createClass(StormM16, [{
        key: "x",
        value: function x(stormM16) {
            if (stormM16 instanceof StormM16) {
                return $M16([this.e[0] * stormM16.e[0] + this.e[1] * stormM16.e[4] + this.e[2] * stormM16.e[8] + this.e[3] * stormM16.e[12], this.e[0] * stormM16.e[1] + this.e[1] * stormM16.e[5] + this.e[2] * stormM16.e[9] + this.e[3] * stormM16.e[13], this.e[0] * stormM16.e[2] + this.e[1] * stormM16.e[6] + this.e[2] * stormM16.e[10] + this.e[3] * stormM16.e[14], this.e[0] * stormM16.e[3] + this.e[1] * stormM16.e[7] + this.e[2] * stormM16.e[11] + this.e[3] * stormM16.e[15], this.e[4] * stormM16.e[0] + this.e[5] * stormM16.e[4] + this.e[6] * stormM16.e[8] + this.e[7] * stormM16.e[12], this.e[4] * stormM16.e[1] + this.e[5] * stormM16.e[5] + this.e[6] * stormM16.e[9] + this.e[7] * stormM16.e[13], this.e[4] * stormM16.e[2] + this.e[5] * stormM16.e[6] + this.e[6] * stormM16.e[10] + this.e[7] * stormM16.e[14], this.e[4] * stormM16.e[3] + this.e[5] * stormM16.e[7] + this.e[6] * stormM16.e[11] + this.e[7] * stormM16.e[15], this.e[8] * stormM16.e[0] + this.e[9] * stormM16.e[4] + this.e[10] * stormM16.e[8] + this.e[11] * stormM16.e[12], this.e[8] * stormM16.e[1] + this.e[9] * stormM16.e[5] + this.e[10] * stormM16.e[9] + this.e[11] * stormM16.e[13], this.e[8] * stormM16.e[2] + this.e[9] * stormM16.e[6] + this.e[10] * stormM16.e[10] + this.e[11] * stormM16.e[14], this.e[8] * stormM16.e[3] + this.e[9] * stormM16.e[7] + this.e[10] * stormM16.e[11] + this.e[11] * stormM16.e[15], this.e[12] * stormM16.e[0] + this.e[13] * stormM16.e[4] + this.e[14] * stormM16.e[8] + this.e[15] * stormM16.e[12], this.e[12] * stormM16.e[1] + this.e[13] * stormM16.e[5] + this.e[14] * stormM16.e[9] + this.e[15] * stormM16.e[13], this.e[12] * stormM16.e[2] + this.e[13] * stormM16.e[6] + this.e[14] * stormM16.e[10] + this.e[15] * stormM16.e[14], this.e[12] * stormM16.e[3] + this.e[13] * stormM16.e[7] + this.e[14] * stormM16.e[11] + this.e[15] * stormM16.e[15]]);
            } else {
                return $M16([0.0, 0.0, 0.0, this.e[0] * stormM16.e[0] + this.e[1] * stormM16.e[1] + this.e[2] * stormM16.e[2], 0.0, 0.0, 0.0, this.e[4] * stormM16.e[0] + this.e[5] * stormM16.e[1] + this.e[6] * stormM16.e[2], 0.0, 0.0, 0.0, this.e[8] * stormM16.e[0] + this.e[9] * stormM16.e[1] + this.e[10] * stormM16.e[2], 0.0, 0.0, 0.0, this.e[12] * stormM16.e[0] + this.e[13] * stormM16.e[1] + this.e[14] * stormM16.e[2]]);
            }
        }
    }, {
        key: "transpose",

        /**
        * Get matrix transposed
        * @returns {StormM16}
        */
        value: function transpose() {
            return $M16([this.e[0], this.e[4], this.e[8], this.e[12], this.e[1], this.e[5], this.e[9], this.e[13], this.e[2], this.e[6], this.e[10], this.e[14], this.e[3], this.e[7], this.e[11], this.e[15]]);
        }
    }, {
        key: "inverse",

        /**
        * Get matrix inverted
        * @returns {StormM16}
        */
        value: function inverse() {
            var src = $M16(this.e).transpose().e;
            var tmp = new Float32Array(12);
            var dst = new Float32Array(16);
            // Calculate pairs for first 8 elements (cofactors)
            tmp[0] = src[10] * src[15];
            tmp[1] = src[11] * src[14];
            tmp[2] = src[9] * src[15];
            tmp[3] = src[11] * src[13];
            tmp[4] = src[9] * src[14];
            tmp[5] = src[10] * src[13];
            tmp[6] = src[8] * src[15];
            tmp[7] = src[11] * src[12];
            tmp[8] = src[8] * src[14];
            tmp[9] = src[10] * src[12];
            tmp[10] = src[8] * src[13];
            tmp[11] = src[9] * src[12];

            // Calculate first 8 elements (cofactors)
            dst[0] = tmp[0] * src[5] + tmp[3] * src[6] + tmp[4] * src[7];
            dst[0] -= tmp[1] * src[5] + tmp[2] * src[6] + tmp[5] * src[7];
            dst[1] = tmp[1] * src[4] + tmp[6] * src[6] + tmp[9] * src[7];
            dst[1] -= tmp[0] * src[4] + tmp[7] * src[6] + tmp[8] * src[7];
            dst[2] = tmp[2] * src[4] + tmp[7] * src[5] + tmp[10] * src[7];
            dst[2] -= tmp[3] * src[4] + tmp[6] * src[5] + tmp[11] * src[7];
            dst[3] = tmp[5] * src[4] + tmp[8] * src[5] + tmp[11] * src[6];
            dst[3] -= tmp[4] * src[4] + tmp[9] * src[5] + tmp[10] * src[6];
            dst[4] = tmp[1] * src[1] + tmp[2] * src[2] + tmp[5] * src[3];
            dst[4] -= tmp[0] * src[1] + tmp[3] * src[2] + tmp[4] * src[3];
            dst[5] = tmp[0] * src[0] + tmp[7] * src[2] + tmp[8] * src[3];
            dst[5] -= tmp[1] * src[0] + tmp[6] * src[2] + tmp[9] * src[3];
            dst[6] = tmp[3] * src[0] + tmp[6] * src[1] + tmp[11] * src[3];
            dst[6] -= tmp[2] * src[0] + tmp[7] * src[1] + tmp[10] * src[3];
            dst[7] = tmp[4] * src[0] + tmp[9] * src[1] + tmp[10] * src[2];
            dst[7] -= tmp[5] * src[0] + tmp[8] * src[1] + tmp[11] * src[2];

            // Calculate pairs for second 8 elements (cofactors)
            tmp[0] = src[2] * src[7];
            tmp[1] = src[3] * src[6];
            tmp[2] = src[1] * src[7];
            tmp[3] = src[3] * src[5];
            tmp[4] = src[1] * src[6];
            tmp[5] = src[2] * src[5];
            tmp[6] = src[0] * src[7];
            tmp[7] = src[3] * src[4];
            tmp[8] = src[0] * src[6];
            tmp[9] = src[2] * src[4];
            tmp[10] = src[0] * src[5];
            tmp[11] = src[1] * src[4];

            // Calculate second 8 elements (cofactors)
            dst[8] = tmp[0] * src[13] + tmp[3] * src[14] + tmp[4] * src[15];
            dst[8] -= tmp[1] * src[13] + tmp[2] * src[14] + tmp[5] * src[15];
            dst[9] = tmp[1] * src[12] + tmp[6] * src[14] + tmp[9] * src[15];
            dst[9] -= tmp[0] * src[12] + tmp[7] * src[14] + tmp[8] * src[15];
            dst[10] = tmp[2] * src[12] + tmp[7] * src[13] + tmp[10] * src[15];
            dst[10] -= tmp[3] * src[12] + tmp[6] * src[13] + tmp[11] * src[15];
            dst[11] = tmp[5] * src[12] + tmp[8] * src[13] + tmp[11] * src[14];
            dst[11] -= tmp[4] * src[12] + tmp[9] * src[13] + tmp[10] * src[14];
            dst[12] = tmp[2] * src[10] + tmp[5] * src[11] + tmp[1] * src[9];
            dst[12] -= tmp[4] * src[11] + tmp[0] * src[9] + tmp[3] * src[10];
            dst[13] = tmp[8] * src[11] + tmp[0] * src[8] + tmp[7] * src[10];
            dst[13] -= tmp[6] * src[10] + tmp[9] * src[11] + tmp[1] * src[8];
            dst[14] = tmp[6] * src[9] + tmp[11] * src[11] + tmp[3] * src[8];
            dst[14] -= tmp[10] * src[11] + tmp[2] * src[8] + tmp[7] * src[9];
            dst[15] = tmp[10] * src[10] + tmp[4] * src[8] + tmp[9] * src[9];
            dst[15] -= tmp[8] * src[9] + tmp[11] * src[10] + tmp[5] * src[8];

            var det = src[0] * dst[0] + src[1] * dst[1] + src[2] * dst[2] + src[3] * dst[3];

            return $M16([dst[0] * det, dst[1] * det, dst[2] * det, dst[3] * det, dst[4] * det, dst[5] * det, dst[6] * det, dst[7] * det, dst[8] * det, dst[9] * det, dst[10] * det, dst[11] * det, dst[12] * det, dst[13] * det, dst[14] * det, dst[15] * det]);
        }
    }, {
        key: "getLeft",

        /**
        * Get vector left
        * @returns {StormV3}
        */
        value: function getLeft() {
            return $V3([this.e[0], this.e[4], this.e[8]]).normalize();
        }
    }, {
        key: "getUp",

        /**
        * Get vector up
        * @returns {StormV3}
        */
        value: function getUp() {
            return $V3([this.e[1], this.e[5], this.e[9]]).normalize();
        }
    }, {
        key: "getForward",

        /**
        * Get vector forward
        * @returns {StormV3}
        */
        value: function getForward() {
            return $V3([this.e[2], this.e[6], this.e[10]]).normalize();
        }
    }, {
        key: "setRotation",

        /**
        * Set rotation in axis
        * @returns {StormM16}
        * @param {number} radians Radians
        * @param {String} [relative=true] false for absolute rotation
        * @param {StormV3} [axis=$V3([0.0,1.0,0.0])]
        */
        value: function setRotation(radians, relative, axis) {
            if (axis !== undefined && axis !== null && axis.e[0] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([1.0, 0.0, 0.0, 0.0, 0.0, Math.cos(radians), -Math.sin(radians), 0.0, 0.0, Math.sin(radians), Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([1.0, 0.0, 0.0, 0.0, 0.0, Math.cos(radians), -Math.sin(radians), 0.0, 0.0, Math.sin(radians), Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
            if (axis !== undefined && axis !== null && axis.e[1] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([Math.cos(radians), 0.0, Math.sin(radians), 0.0, 0.0, 1.0, 0.0, 0.0, -Math.sin(radians), 0.0, Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([Math.cos(radians), 0.0, Math.sin(radians), 0.0, 0.0, 1.0, 0.0, 0.0, -Math.sin(radians), 0.0, Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
            if (axis !== undefined && axis !== null && axis.e[2] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([Math.cos(radians), -Math.sin(radians), 0.0, 0.0, Math.sin(radians), Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([Math.cos(radians), -Math.sin(radians), 0.0, 0.0, Math.sin(radians), Math.cos(radians), 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
        }
    }, {
        key: "setRotationX",

        /**
        * Set rotation x
        * @returns {StormM16}
        * @param {number} radians Radians
        * @param {String} [relative=true] false for absolute rotation
        */
        value: function setRotationX(radians, relative) {
            return this.setRotation(radians, relative, $V3([1.0, 0.0, 0.0]));
        }
    }, {
        key: "setRotationY",

        /**
        * Set rotation y
        * @returns {StormM16}
        * @param {number} radians Radians
        * @param {String} [relative=true] false for absolute rotation
        */
        value: function setRotationY(radians, relative) {
            return this.setRotation(radians, relative, $V3([0.0, 1.0, 0.0]));
        }
    }, {
        key: "setRotationZ",

        /**
        * Set rotation z
        * @returns {StormM16}
        * @param {number} radians Radians
        * @param {String}[relative=true] false for absolute rotation
        */
        value: function setRotationZ(radians, relative) {
            return this.setRotation(radians, relative, $V3([0.0, 0.0, 1.0]));
        }
    }, {
        key: "setScale",

        /**
        * Set scale in axis
        * @returns {StormM16}
        * @param {number} scale Scale
        * @param {String} [relative=true] false for absolute scale
        * @param {StormV3} [axis=$V3([0.0,1.0,0.0])]
        */
        value: function setScale(scale, relative, axis) {
            if (axis !== undefined && axis !== null && axis.e[0] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([scale, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([scale, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
            if (axis !== undefined && axis !== null && axis.e[1] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([1.0, 0.0, 0.0, 0.0, 0.0, scale, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([1.0, 0.0, 0.0, 0.0, 0.0, scale, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
            if (axis !== undefined && axis !== null && axis.e[2] !== null) {
                if (relative === undefined || relative === null || relative === true) return this.x($M16([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, scale, 0.0, 0.0, 0.0, 0.0, 1.0]));else return $M16([1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, scale, 0.0, 0.0, 0.0, 0.0, 1.0]);
            }
        }
    }, {
        key: "setScaleX",

        /**
        * Set scale x
        * @returns {StormM16}
        * @param {number} scale Scale
        * @param {String} [relative=true] false for absolute scale
        */
        value: function setScaleX(scale, relative) {
            return this.setScale(scale, relative, $V3([1.0, 0.0, 0.0]));
        }
    }, {
        key: "setScaleY",

        /**
        * Set scale y
        * @returns {StormM16}
        * @param {number} scale Scale
        * @param {String} [relative=true] false for absolute scale
        */
        value: function setScaleY(scale, relative) {
            return this.setScale(scale, relative, $V3([0.0, 1.0, 0.0]));
        }
    }, {
        key: "setScaleZ",

        /**
        * Set scale z
        * @returns {StormM16}
        * @param {number} scale Scale
        * @param {String}[relative=true] false for absolute scale
        */
        value: function setScaleZ(scale, relative) {
            return this.setScale(scale, relative, $V3([0.0, 0.0, 1.0]));
        }
    }, {
        key: "getPosition",

        /**
        * Get the position vector
        * @returns {StormV3}
        */
        value: function getPosition() {
            // return V3
            return $V3([this.e[3], this.e[7], this.e[11]]);
        }
    }, {
        key: "setPosition",

        /**
        * Set the position vector
        * @returns {StormM16}
        * @param {StormV3} vec Position vector
        */
        value: function setPosition(vec) {
            // return M16
            this.e[3] = vec.e[0];
            this.e[7] = vec.e[1];
            this.e[11] = vec.e[2];
            return this;
        }
    }], [{
        key: "makeLookAt",

        /**
        * Look at
        * @returns {StormM16}
        * @param {number} ex Eye x
        * @param {number} ey Eye y
        * @param {number} ez Eye z
        * @param {number} cx Center x
        * @param {number} cy Center y
        * @param {number} cz Center z
        * @param {number} ux Up axis x
        * @param {number} uy Up axis y
        * @param {number} uz Up axis z
        */
        value: function makeLookAt(ex, ey, ez, cx, cy, cz, ux, uy, uz) {
            var eye = $V3([ex, ey, ez]);
            var center = $V3([cx, cy, cz]);
            var up = $V3([ux, uy, uz]);

            var mag = void 0;

            var z = eye.subtract(center).normalize();
            var x = up.cross(z).normalize();
            var y = z.cross(x).normalize();

            var m = $M16([x.e[0], x.e[1], x.e[2], 0, y.e[0], y.e[1], y.e[2], 0, z.e[0], z.e[1], z.e[2], 0, 0, 0, 0, 1]);

            var t = $M16([1, 0, 0, -ex, 0, 1, 0, -ey, 0, 0, 1, -ez, 0, 0, 0, 1]);
            return m.x(t);
        }
    }, {
        key: "setPerspectiveProjection",
        value: function setPerspectiveProjection(fovy, aspect, znear, zfar) {
            var top = Math.tan(fovy * Math.PI / 360) * znear;
            var bottom = -top;
            var left = aspect * bottom;
            var right = aspect * top;

            var X = 2 * znear / (right - left);
            var Y = 2 * znear / (top - bottom);
            var A = (right + left) / (right - left);
            var B = (top + bottom) / (top - bottom);
            var C = -(zfar + znear) / (zfar - znear);
            var D = -2 * zfar * znear / (zfar - znear);

            return $M16([X, 0, A, 0, 0, Y, B, 0, 0, 0, C, D, 0, 0, -1, 0]);
        }
    }, {
        key: "setOrthographicProjection",
        value: function setOrthographicProjection(left, right, bottom, top, znear, zfar) {
            var tx = -(right + left) / (right - left);
            var ty = -(top + bottom) / (top - bottom);
            var tz = -(zfar + znear) / (zfar - znear);

            return $M16([2 / (right - left), 0, 0, tx, 0, 2 / (top - bottom), 0, ty, 0, 0, -2 / (zfar - znear), tz, 0, 0, 0, 1]);
        }
    }]);

    return StormM16;
}();
/**
 * @returns {StormM16}
 * @param {Array<number>|Float32Array} elements Array with 16 values
 */


function $M16(elements) {
    return new StormM16(elements);
}
global.StormM16 = StormM16;
module.exports.StormM16 = StormM16;

global.$M16 = $M16;
module.exports.$M16 = $M16;

/**
* @class
* @constructor

* @property {Float32Array} e Array with 3 values
*/

var StormV3 = exports.StormV3 = function () {
    function StormV3(elements) {
        _classCallCheck(this, StormV3);

        this.e = new Float32Array(elements);
    }

    /**
    * Perform a + b and get the vector
    * @returns {StormV3}
    * @param {StormV3} vector Vector b
    */


    _createClass(StormV3, [{
        key: "add",
        value: function add(vector) {
            return $V3([this.e[0] + vector.e[0], this.e[1] + vector.e[1], this.e[2] + vector.e[2]]);
        }
    }, {
        key: "cross",

        /**
        * Perform a cross b and get the vector
        * @returns {StormV3}
        * @param {StormV3} vector Vector b
        */
        value: function cross(vector) {
            return $V3([this.e[1] * vector.e[2] - this.e[2] * vector.e[1], this.e[2] * vector.e[0] - this.e[0] * vector.e[2], this.e[0] * vector.e[1] - this.e[1] * vector.e[0]]);
        }
    }, {
        key: "distance",

        /**
        * Get the distance between a and b
        * @returns {number}
        * @param {StormV3} vector Vector b
        */
        value: function distance(vector) {
            return Math.sqrt((this.e[0] - vector.e[0]) * (this.e[0] - vector.e[0]) + (this.e[1] - vector.e[1]) * (this.e[1] - vector.e[1]) + (this.e[2] - vector.e[2]) * (this.e[2] - vector.e[2]));
        }
    }, {
        key: "dot",

        /**
        * Get the dot product between a and b
        * @returns {number}
        * @param {StormV3} vector Vector b
        */
        value: function dot(vector) {
            return this.e[0] * vector.e[0] + this.e[1] * vector.e[1] + this.e[2] * vector.e[2];
        }
    }, {
        key: "equal",

        /**
        * Check if a and b have the same values
        * @returns {boolean}
        * @param {StormV3} vector Vector b
        */
        value: function equal(vector) {
            return this.e[0] === vector.e[0] && this.e[1] === vector.e[1] && this.e[2] === vector.e[2];
        }
    }, {
        key: "modulus",

        /**
        * Get the length of the vector
        * @returns {number}
        */
        value: function modulus() {
            return Math.sqrt(this.sumComponentSqrs());
        }
    }, {
        key: "sumComponentSqrs",
        value: function sumComponentSqrs() {
            var V3 = this.sqrComponents();

            return V3[0] + V3[1] + V3[2];
        }
    }, {
        key: "sqrComponents",
        value: function sqrComponents() {
            var V3 = new Float32Array(3);
            V3.set([this.e[0] * this.e[0], this.e[1] * this.e[1], this.e[2] * this.e[2]]);

            return V3;
        }
    }, {
        key: "x",

        /**
        * Perform a * b and get the vector
        * @returns {StormV3|StormM16}
        * @param {StormV3|StormM16|number} value Value b (Vector|Matrix|Float)
        */
        value: function x(value) {
            var typeVector = value instanceof StormV3;
            var typeMatrix = value instanceof StormM16;
            if (typeVector) {
                return $V3([this.e[0] * value.e[0], this.e[1] * value.e[1], this.e[2] * value.e[2]]);
            } else if (typeMatrix) {
                var thisV3 = $M16([1.0, 0.0, 0.0, this.e[0], 0.0, 1.0, 0.0, this.e[1], 0.0, 0.0, 1.0, this.e[2], 0.0, 0.0, 0.0, 1.0]);
                return $M16([thisV3.e[0] + 0.0 + 0.0 + 0.0, 0.0 + thisV3.e[1] + 0.0 + 0.0, 0.0 + 0.0 + thisV3.e[2] + 0.0, thisV3.e[0] * value.e[0] + thisV3.e[1] * value.e[1] + thisV3.e[2] * value.e[2] + thisV3.e[3], thisV3.e[4] + 0.0 + 0.0 + 0.0, 0.0 + thisV3.e[5] + 0.0 + 0.0, 0.0 + 0.0 + thisV3.e[6] + 0.0, thisV3.e[4] * value.e[0] + thisV3.e[5] * value.e[1] + thisV3.e[6] * value.e[2] + thisV3.e[7], thisV3.e[8] + 0.0 + 0.0 + 0.0, 0.0 + thisV3.e[9] + 0.0 + 0.0, 0.0 + 0.0 + thisV3.e[10] + 0.0, thisV3.e[8] * value.e[0] + thisV3.e[9] * value.e[1] + thisV3.e[10] * value.e[2] + thisV3.e[11], thisV3.e[12] + 0.0 + 0.0 + 0.0, 0.0 + thisV3.e[13] + 0.0 + 0.0, 0.0 + 0.0 + thisV3.e[14] + 0.0, thisV3.e[12] * value.e[0] + thisV3.e[13] * value.e[1] + thisV3.e[14] * value.e[2] + thisV3.e[15] * value.e[15]]);
            } else return $V3([this.e[0] * value, this.e[1] * value, this.e[2] * value]);
        }
    }, {
        key: "reflect",

        /**
        * Get the reflected vector with b
        * @returns {StormV3}
        * @param {StormV3} vector Vector b
        */
        value: function reflect(vector) {
            var I = $V3([this.e[0], this.e[1], this.e[2]]);
            var N = $V3([vector.e[0], vector.e[1], vector.e[2]]);

            var vec = N.x(I.dot(N));
            vec = vec.x(2.0);
            vec = I.subtract(vec);

            return vec;
        }
    }, {
        key: "subtract",

        /**
        * Perform a - b and get the vector
        * @returns {StormV3}
        * @param {StormV3} vector Vector b
        */
        value: function subtract(vector) {
            return $V3([this.e[0] - vector.e[0], this.e[1] - vector.e[1], this.e[2] - vector.e[2]]);
        }
    }, {
        key: "normalize",

        /**
        * Get the vector normalized
        * @returns {StormV3}
        */
        value: function normalize() {
            var inverse = 1.0 / this.modulus();

            return $V3([this.e[0] * inverse, this.e[1] * inverse, this.e[2] * inverse]);
        }
    }]);

    return StormV3;
}();
/**
* @returns {StormV3}
* @param {Array<number>} elements Array with 3 values
*/


function $V3(elements) {
    return new StormV3(elements);
}
global.StormV3 = StormV3;
module.exports.StormV3 = StormV3;

global.$V3 = $V3;
module.exports.$V3 = $V3;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],6:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Utils = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

require("webclgl");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
* @class
*/
var Utils = exports.Utils = function () {
    function Utils() {
        _classCallCheck(this, Utils);

        window.requestAnimFrame = function () {
            return window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || window.oRequestAnimationFrame || window.msRequestAnimationFrame || function (callback) {
                window.setTimeout(callback, 1000 / 60);
            };
        }();
    }

    /**
    * Get HTMLCanvasElement from Uint8Array
    * @param {Uint8Array} uint8arr
    * @param {int} width
    * @param {int} height
     * @returns {HTMLCanvasElement}
    */


    _createClass(Utils, [{
        key: "getImageFromCanvas",

        /**
         * @callback Utils~getImageFromCanvas~onload
         * @param {HTMLImageElement} img
         */
        /**
        * Get HTMLImageElement from canvas
        * @param {HTMLCanvasElement} oldCanvas
        * @param {Utils~getImageFromCanvas~onload} onload
        */
        value: function getImageFromCanvas(oldCanvas, onload) {
            var imagen = document.createElement('img');
            imagen.onload = function () {
                onload(imagen);
            };
            imagen.src = oldCanvas.toDataURL();
        }
    }, {
        key: "pack",


        /**
         * Angle between two vectors viewing from top
         * @returns {number}
         * @param {StormV3} vectorA
         * @param {StormV3} vectorB
          Utils.prototype.angle = function(vA, vB) {
            let vAA = vA.normalize();
            let vBB = vB.normalize();
              let escalarProduct = Math.acos((vAA.e[0]*vBB.e[0])+(vAA.e[1]*vBB.e[1])+(vAA.e[2]*vBB.e[2]));
              let vCC = vAA.cross(vBB);
            //console.log(vCC.e[0]+" "+vCC.e[1]+" "+vCC.e[2]);
              if(vCC.e[1] == 1) {
                escalarProduct = (Math.PI+escalarProduct);
            }
              return escalarProduct;
        };*/

        /**
        * Pack 1float (0.0-1.0) to 4float rgba (0.0-1.0, 0.0-1.0, 0.0-1.0, 0.0-1.0)*
        * @param {number} value
        * @returns {Array<number>}
        *
        */
        value: function pack(value) {
            var bias = [1.0 / 255.0, 1.0 / 255.0, 1.0 / 255.0, 0.0];

            var r = value;
            var g = Utils.fract(r * 255.0);
            var b = Utils.fract(g * 255.0);
            var a = Utils.fract(b * 255.0);
            var colour = [r, g, b, a];

            var dd = [colour[1] * bias[0], colour[2] * bias[1], colour[3] * bias[2], colour[3] * bias[3]];

            return [colour[0] - dd[0], colour[1] - dd[1], colour[2] - dd[2], colour[3] - dd[3]];
        }
    }, {
        key: "unpack",

        /**
        * Unpack 4float rgba (0.0-1.0, 0.0-1.0, 0.0-1.0, 0.0-1.0) to 1float (0.0-1.0)
        * @param {Array<number>} value
         * @returns {number}
        */
        value: function unpack(value) {
            var bitShifts = [1.0, 1.0 / 255.0, 1.0 / (255.0 * 255.0), 1.0 / (255.0 * 255.0 * 255.0)];
            return Utils.dot4(value, bitShifts);
        }
    }], [{
        key: "getCanvasFromUint8Array",
        value: function getCanvasFromUint8Array(uint8arr, width, height) {
            var e = document.createElement('canvas');
            e.width = width;
            e.height = height;
            var ctx2D = e.getContext("2d");
            var image = ctx2D.createImageData(width, height);
            for (var i = 0; i < image.data.length; i++) {
                image.data[i] = uint8arr[i];
            }ctx2D.putImageData(image, 0, 0);

            return e;
        }
    }, {
        key: "getUint8ArrayFromHTMLImageElement",


        /**
        * Get Uint8Array from HTMLImageElement
        * @param {HTMLImageElement} imageElement
         * @returns {Uint8ClampedArray}
        */
        value: function getUint8ArrayFromHTMLImageElement(imageElement) {
            var e = document.createElement('canvas');
            e.width = imageElement.width;
            e.height = imageElement.height;
            var ctx2D_tex = e.getContext("2d");
            ctx2D_tex.drawImage(imageElement, 0, 0);
            var arrayTex = ctx2D_tex.getImageData(0, 0, imageElement.width, imageElement.height);

            return arrayTex.data;
        }
    }, {
        key: "getVector",


        /**
        * Get random vector from vecNormal with deviation in degrees
        * @param {StormV3} vecNormal
        * @param {number} degrees
         * @returns {StormV3}
        */
        value: function getVector(vecNormal, degrees) {
            var ob = Utils.cartesianToSpherical(vecNormal);
            var angleLat = ob.lat;
            var angleLng = ob.lng;

            var desvLat = Math.random() * 180.0 - 90.0;
            var desvLng = Math.random() * 180.0 - 90.0;
            angleLat += degrees * desvLat;
            angleLng += degrees * desvLng;

            return Utils.sphericalToCartesian(1.0, angleLat, angleLng);
        }
    }, {
        key: "getVectorGLSLFunctionString",

        /**
        * Get random vector from vecNormal with deviation in degrees (GLSL)
        * @returns {String}
        */
        value: function getVectorGLSLFunctionString() {
            return 'vec3 getVector(vec3 vecNormal, float degrees, vec2 vecNoise) {\n' + 'vec3 ob = cartesianToSpherical(vecNormal);' + 'float angleLat = ob.y;' + 'float angleLng = ob.z;' + 'float desvLat = (vecNoise.x*180.0)-90.0;' + 'float desvLng = (vecNoise.y*180.0)-90.0;' + 'angleLat += (degrees*desvLat);' + 'angleLng += (degrees*desvLng);' + 'return sphericalToCartesian(vec3(1.0, angleLat, angleLng));' + '}\n';
        }
    }, {
        key: "cartesianToSpherical",


        /**
         * cartesianToSpherical
         * @param {StormV3} vec
         * @returns {Object}
         * @example
         * $V3([1,0,0])  return {radius: 1, lat: 90, lng: 0}
         * $V3([0,0,1])  return {radius: 1, lat: 90, lng: 90}
         * $V3([-1,0,0]) return {radius: 1, lat: 90, lng: 180}
         * $V3([0,0,-1]) return {radius: 1, lat: 90, lng: -90}
         */
        value: function cartesianToSpherical(vec) {
            var r = Math.sqrt(vec.e[0] * vec.e[0] + vec.e[1] * vec.e[1] + vec.e[2] * vec.e[2]);

            var angleLat = Utils.radToDeg(Math.acos(vec.e[1] / r));
            var angleLng = Utils.radToDeg(Math.atan2(vec.e[2], vec.e[0]));

            return { "radius": r,
                "lat": angleLat,
                "lng": angleLng };
        }
    }, {
        key: "cartesianToSphericalGLSLFunctionString",

        /**
         * cartesianToSpherical (GLSL)
        * @returns {String}
        */
        value: function cartesianToSphericalGLSLFunctionString() {
            return 'vec3 cartesianToSpherical(vec3 vect) {\n' + 'float r = sqrt(vect.x*vect.x + vect.y*vect.y + vect.z*vect.z);' + 'float angleLat = radToDeg(acos(vect.y/r));' + 'float angleLng = radToDeg(atan(vect.z, vect.x));' + 'return vec3(r, angleLat, angleLng);' + '}\n';
        }
    }, {
        key: "sphericalToCartesian",


        /**
         * sphericalToCartesian
         * @param {number} radius
         * @param {number} lat Lat in degrees
         * @param {number} lng Lng in degrees
         * @returns {StormV3}
         * @example
         * (1.0, 90.0, 0.0).e) return $V3([1,0,0])
         * (1.0, 90.0, 90.0).e) return $V3([0,0,1])
         * (1.0, 90.0, 180.0).e) return $V3([-1,0,0])
         * (1.0, 90.0, -90.0).e) return $V3([0,0,-1])
         **/
        value: function sphericalToCartesian(radius, lat, lng) {
            var r = radius;
            var angleLat = Utils.degToRad(lat);
            var angleLng = Utils.degToRad(lng);

            var x = r * Math.sin(angleLat) * Math.cos(angleLng);
            var z = r * Math.sin(angleLat) * Math.sin(angleLng);
            var y = r * Math.cos(angleLat);

            return new StormV3([x, y, z]);
        }
    }, {
        key: "sphericalToCartesianGLSLFunctionString",

        /**
         * sphericalToCartesian (GLSL)
        * @returns {String}
        */
        value: function sphericalToCartesianGLSLFunctionString() {
            return 'vec3 sphericalToCartesian(vec3 vect) {\n' + 'float r = vect.x;' + 'float angleLat = degToRad(vect.y);' + 'float angleLng = degToRad(vect.z);' + 'float x = r*sin(angleLat)*cos(angleLng);' + 'float z = r*sin(angleLat)*sin(angleLng);' + 'float y = r*cos(angleLat);' + 'return vec3(x,y,z);' + '}\n';
        }
    }, {
        key: "refract",


        /**
        * Refract
        * @param {StormV3} V
        * @param {StormV3} N
        * @param {number} n1 Refract index way 1
        * @param {number} n2 Refract index way 2
         * @returns {StormV3}
        */
        value: function refract(V, N, n1, n2) {
            var refrIndex = n1 / n2;
            var cosI = N.dot(V) * -1.0;
            var cosT2 = 1.0 - refrIndex * refrIndex * (1.0 - cosI * cosI);
            var vv = V.x(refrIndex);
            return vv.add(N.x(refrIndex * cosI - Math.sqrt(cosT2)));
        }
    }, {
        key: "degToRad",


        /**
        * Degrees to radians. Full circle = 360 degrees.
        * @param {number} degrees
         * @returns {number}
        */
        value: function degToRad(degrees) {
            return degrees * 3.14159 / 180;
        }
    }, {
        key: "degToRadGLSLFunctionString",

        /**
         * Degrees to radians. Full circle = 360 degrees. (GLSL)
        * @returns {String}
        */
        value: function degToRadGLSLFunctionString() {
            return 'float degToRad(float deg) {' + 'return (deg*3.14159)/180.0;' + '}';
        }
    }, {
        key: "radToDeg",


        /**
        * Radians to degrees
        * @param {number} radians
         * @returns {number}
        */
        value: function radToDeg(radians) {
            return radians * (180 / 3.14159);
        }
    }, {
        key: "radToDegGLSLFunctionString",

        /**
         * Radians to degrees (GLSL)
        * @returns {String}
        */
        value: function radToDegGLSLFunctionString() {
            return 'float radToDeg(float rad) {' + 'return rad*(180.0/3.14159);' + '}';
        }
    }, {
        key: "hexToRgb",


        /**
         * hexToRgb
         * @param {String} hex
         * @returns  {Array<number>} rgb values from 0 to 255
         */
        value: function hexToRgb(hex) {
            // Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
            var shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
            hex = hex.replace(shorthandRegex, function (m, r, g, b) {
                return r + r + g + g + b + b;
            });

            var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
            return result ? {
                r: parseInt(result[1], 16),
                g: parseInt(result[2], 16),
                b: parseInt(result[3], 16)
            } : null;
        }
    }, {
        key: "rgbToHex",

        /**
         * rgbToHex
         * @param {Array<number>} rgb values from 0 to 255
         * @returns {String}
         */
        value: function rgbToHex(rgb) {
            var rgbVal = rgb[2] | rgb[1] << 8 | rgb[0] << 16;
            return '#' + (0x1000000 + rgbVal).toString(16).slice(1);
        }
    }, {
        key: "invsqrt",


        /**
        * Inverse sqrt
        * @param {number} value
         * @returns {number}
        */
        value: function invsqrt(value) {
            return 1.0 / value;
        }
    }, {
        key: "smoothstep",


        /**
        * Smoothstep
        * @param {number} edge0
        * @param {number} edge1
        * @param {number} current
         * @returns {number}
        */
        value: function smoothstep(edge0, edge1, current) {
            if (current < edge0) return 0;
            if (current >= edge1) return 1;
            if (edge0 === edge1) return -1;
            var p = (current - edge0) / (edge1 - edge0);

            return p * p * (3 - 2 * p);
        }
    }, {
        key: "dot4",


        /**
        * Dot product vector4float
        * @param {Array<number>} vector4A Vector a
        * @param {Array<number>} vector4B Vector b
         * @returns {number}
        */
        value: function dot4(vector4A, vector4B) {
            return vector4A[0] * vector4B[0] + vector4A[1] * vector4B[1] + vector4A[2] * vector4B[2] + vector4A[3] * vector4B[3];
        }
    }, {
        key: "fract",


        /**
        * Compute the fractional part of the argument. Example: fract(pi)=0.14159265...
        * @param {number} value
         * @returns {number}
        */
        value: function fract(value) {
            return value > 0 ? value - Math.floor(value) : value - Math.ceil(value);
        }
    }, {
        key: "packGLSLFunctionString",

        /**
        * Get pack GLSL function string
        * @returns {String}
        */
        value: function packGLSLFunctionString() {
            return 'vec4 pack (float depth) {' + 'const vec4 bias = vec4( 1.0 / 255.0,' + '1.0 / 255.0,' + '1.0 / 255.0,' + '0.0);' + 'float r = depth;' + 'float g = fract(r * 255.0);' + 'float b = fract(g * 255.0);' + 'float a = fract(b * 255.0);' + 'vec4 colour = vec4(r, g, b, a);' + 'return colour - (colour.yzww * bias);' + '}';
        }
    }, {
        key: "unpackGLSLFunctionString",

        /**
        * Get unpack GLSL function string
        * @returns {String}
        */
        value: function unpackGLSLFunctionString() {
            return 'float unpack (vec4 colour) {' + 'const vec4 bitShifts = vec4(1.0,' + '1.0 / 255.0,' + '1.0 / (255.0 * 255.0),' + '1.0 / (255.0 * 255.0 * 255.0));' + 'return dot(colour, bitShifts);' + '}';
        }
    }, {
        key: "rayTraversalInitSTR",
        value: function rayTraversalInitSTR() {
            return '' + 'float wh = ceil(sqrt(uResolution*uResolution*uResolution));\n' + 'float cs = uGridsize/uResolution;\n' + // cell size
            'float chs = cs/2.0;\n' + // cell size
            'float texelSize = 1.0/(wh-1.0);\n' + // 1.0/(wh-1.0)??

            // Fast Voxel Traversal Algorithm for Ray Tracing. John Amanatides & Andrew Woo.
            // http://www.cse.chalmers.se/edu/course/TDA361/grid.pdf
            // More info:
            // http://www.clockworkcoders.com/oglsl/rt/gpurt3.htm
            'vec3 gl = vec3(-(uGridsize/2.0), -(uGridsize/2.0), -(uGridsize/2.0));\n' + 'vec3 _r = vec3(uGridsize, uGridsize, uGridsize);\n' + 'vec3 _rRes = vec3(uResolution, uResolution, uResolution);\n' + 'vec3 _len = _r/_rRes;\n' + 'vec3 worldToVoxel(vec3 world) {\n' + 'vec3 ijk = (world - gl) / _len;\n' + // (1.0-(-1.0)) / (2/64) = 64
            'ijk = vec3(floor(ijk.x), floor(ijk.y), floor(ijk.z));\n' + 'return ijk;\n' + '}\n' + 'float voxelToWorldX(float x) {return x * _len.x + gl.x;}\n' + // 64*(2/64)+(-1.0) = 1.0
            'float voxelToWorldY(float y) {return y * _len.y + gl.y;}\n' + 'float voxelToWorldZ(float z) {return z * _len.z + gl.z;}\n';
        }
    }, {
        key: "rayTraversalSTR",
        value: function rayTraversalSTR(resolution) {
            return '' + 'vec2 getId(vec3 voxel) {\n' + 'int tex3dId = (int(voxel.y)*(int(uResolution)*int(uResolution)))+(int(voxel.z)*(int(uResolution)))+int(voxel.x);\n' + 'float num = float(tex3dId)/wh;\n' + 'float col = fract(num)*wh;\n' + 'float row = floor(num);\n' + 'return vec2(col*texelSize, row*texelSize);\n' + '}\n' + 'vec4 getVoxel_Color(vec2 texVec, vec3 voxel, vec3 RayOrigin) {\n' + 'vec4 rgba = vec4(0.0,0.0,0.0,0.0);\n' + 'vec4 texture = sampler_voxelColor[vec2(texVec.x, texVec.y)];\n' + 'if(texture.a/255.0 > 0.5) {\n' + // existen tringulos dentro?
            'rgba = vec4(texture.rgb/255.0,distance(vec3(voxelToWorldX(voxel.x), voxelToWorldX(voxel.y), voxelToWorldX(voxel.z)),RayOrigin));\n' + '}\n' + 'return rgba;\n' + '}\n' + 'vec4 getVoxel_Pos(vec2 texVec) {\n' + 'vec4 rgba = vec4(0.0,0.0,0.0,0.0);\n' + 'vec4 texture = sampler_voxelPos[vec2(texVec.x, texVec.y)];\n' +
            //distance(vec3(voxelToWorldX(voxel.x), voxelToWorldX(voxel.y), voxelToWorldX(voxel.z)),RayOrigin)
            'rgba = vec4( ((texture.xyz/255.0)*uGridsize)-(uGridsize/2.0), 1.0);\n' + 'return rgba;\n' + '}\n' + 'vec4 getVoxel_Normal(vec2 texVec) {\n' + 'vec4 rgba = vec4(0.0,0.0,0.0,0.0);\n' + 'vec4 texture = sampler_voxelNormal[vec2(texVec.x, texVec.y)];\n' + 'rgba = vec4(((texture.rgb/255.0)*2.0)-1.0, 1.0);\n' + 'return rgba;\n' + '}\n' + 'struct RayTraversalResponse {' + 'vec4 voxelColor;' + 'vec4 voxelPos;' + 'vec4 voxelNormal;' + '};' + 'RayTraversalResponse rayTraversal(vec3 RayOrigin, vec3 RayDir) {\n' + 'vec4 fvoxelColor = vec4(0.0, 0.0, 0.0, 0.0);' + 'vec4 fvoxelPos = vec4(0.0, 0.0, 0.0, 0.0);' + 'vec4 fvoxelNormal = vec4(0.0, 0.0, 0.0, 0.0);' + 'vec3 voxel = worldToVoxel(RayOrigin);' + 'vec3 _dir = normalize(RayDir);' + 'vec3 tMax;' + 'if(RayDir.x < 0.0) tMax.x = (voxelToWorldX(voxel.x)-RayOrigin.x)/RayDir.x;' + 'if(RayDir.x > 0.0) tMax.x = (voxelToWorldX(voxel.x+1.0)-RayOrigin.x)/RayDir.x;' + 'if(RayDir.y < 0.0) tMax.y = (voxelToWorldY(voxel.y)-RayOrigin.y)/RayDir.y;' + 'if(RayDir.y < 0.0) tMax.y = (voxelToWorldY(voxel.y+1.0)-RayOrigin.y)/RayDir.y;' + 'if(RayDir.z < 0.0) tMax.z = (voxelToWorldZ(voxel.z)-RayOrigin.z)/RayDir.z;' + 'if(RayDir.z < 0.0) tMax.z = (voxelToWorldZ(voxel.z+1.0)-RayOrigin.z)/RayDir.z;' + 'float tDeltaX = _r.x/abs(RayDir.x);' + // hasta qu punto se debe avanzar en la direccin del rayo antes de que nos encontramos con un nuevo voxel en la direccin x
            'float tDeltaY = _r.y/abs(RayDir.y);' + 'float tDeltaZ = _r.z/abs(RayDir.z);' + 'float stepX = 1.0; float stepY = 1.0; float stepZ = 1.0;\n' + 'float outX = _r.x; float outY = _r.y; float outZ = _r.z;\n' + 'if(RayDir.x < 0.0) {stepX = -1.0; outX = -1.0;}' + 'if(RayDir.y < 0.0) {stepY = -1.0; outY = -1.0;}' + 'if(RayDir.z < 0.0) {stepZ = -1.0; outZ = -1.0;}' + 'vec4 color = vec4(0.0,0.0,0.0,0.0);\n' + 'bool c1; bool c2; bool c3; bool isOut;' + 'vec2 vid;' + 'for(int c = 0; c < ' + resolution + '*2; c++) {\n' + 'c1 = bool(tMax.x < tMax.y);' + 'c2 = bool(tMax.x < tMax.z);' + 'c3 = bool(tMax.y < tMax.z);' + 'isOut = false;' + 'if (c1 && c2) {' + 'voxel.x += stepX;' + 'if(voxel.x==outX) isOut=true;' + 'tMax.x += tDeltaX;' + '} else if(( (c1 && !c2) || (!c1 && !c3) )) {' + 'voxel.z += stepZ;' + 'if(voxel.z==outZ) isOut=true;' + 'tMax.z += tDeltaZ;' + '} else if(!c1 && c3) {' + 'voxel.y += stepY;' + 'if(voxel.y==outY) isOut=true;' + 'tMax.y += tDeltaY;' + '}' + 'if(isOut == true) break;\n' + 'else {' + 'if((voxel.x >= 0.0 && voxel.x <= _rRes.x && voxel.y >= 0.0 && voxel.y <= _rRes.y && voxel.z >= 0.0 && voxel.z <= _rRes.z)) {;\n' + 'vid = getId(voxel);' + 'vec4 vcc = getVoxel_Color(vid, voxel, RayOrigin);' + 'if(vcc.a != 0.0) {' + 'fvoxelColor = vcc;' + 'break;\n' + '}' + '}' + '}' + '}' + 'fvoxelPos = getVoxel_Pos(vid);' + 'fvoxelNormal = getVoxel_Normal(vid);' + 'return RayTraversalResponse(fvoxelColor, fvoxelPos, fvoxelNormal);' + '}\n';
        }
    }, {
        key: "isPowerOfTwo",
        value: function isPowerOfTwo(x) {
            return (x & x - 1) === 0;
        }
    }, {
        key: "nextHighestPowerOfTwo",
        value: function nextHighestPowerOfTwo(x) {
            --x;
            for (var i = 1; i < 32; i <<= 1) {
                x = x | x >> i;
            }
            return x + 1;
        }
    }, {
        key: "getElementPosition",
        value: function getElementPosition(element) {
            var elem = element,
                tagname = "",
                x = 0,
                y = 0;

            while ((typeof elem === "undefined" ? "undefined" : _typeof(elem)) === "object" && typeof elem.tagName !== "undefined") {
                y += elem.offsetTop;
                x += elem.offsetLeft;
                tagname = elem.tagName.toUpperCase();

                if (tagname === "BODY") elem = 0;

                if ((typeof elem === "undefined" ? "undefined" : _typeof(elem)) === "object") {
                    if (_typeof(elem.offsetParent) === "object") elem = elem.offsetParent;
                }
            }

            return { x: x, y: y };
        }
    }, {
        key: "getWebGLContextFromCanvas",
        value: function getWebGLContextFromCanvas(canvas, ctxOpt) {
            return WebCLGLUtils.getWebGLContextFromCanvas(canvas, ctxOpt);
        }
    }, {
        key: "fullScreen",
        value: function fullScreen() {
            if (!document.fullscreenElement && // alternative standard method
            !document.mozFullScreenElement && !document.webkitFullscreenElement) {
                // current working methods
                if (document.documentElement.requestFullscreen) {
                    document.documentElement.requestFullscreen();
                } else if (document.documentElement.mozRequestFullScreen) {
                    document.documentElement.mozRequestFullScreen();
                } else if (document.documentElement.webkitRequestFullscreen) {
                    document.documentElement.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
                }
            } else {
                if (document.cancelFullScreen) {
                    document.cancelFullScreen();
                } else if (document.mozCancelFullScreen) {
                    document.mozCancelFullScreen();
                } else if (document.webkitCancelFullScreen) {
                    document.webkitCancelFullScreen();
                }
            }
        }
    }]);

    return Utils;
}();

global.Utils = Utils;
module.exports.Utils = Utils;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"webclgl":1}]},{},[3]);
