// scejs/dist/scejs/Node.class.js
(function(){function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}return e})()({1:[function(require,module,exports){
(function (global){
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
* @class
* @constructor
*/
var Node = exports.Node = function () {
	function Node() {
		_classCallCheck(this, Node);

		this._components = {};
		this._name = null;
		this._gl = null;

		this._enabled = true;

		this.onTick = null;
	}

	/**
  * initialize
  * @param {String} name.
  * @param {WebGLRenderingContext} glCtx.
  */


	_createClass(Node, [{
		key: "initialize",
		value: function initialize(name, glCtx) {
			this._name = name;
			this._gl = glCtx;
		}
	}, {
		key: "addComponent",


		/**
   * addComponent
   * @param {Component} component
   */
		value: function addComponent(component) {
			this._components[component.type] = component;
			this._components[component.type].initialize(this, this._gl);
		}
	}, {
		key: "getComponent",


		/**
   * getComponent
   * @param {Constants.COMPONENT_TYPES|int} type
   * @returns {Component}
   */
		value: function getComponent(type) {
			return this._components[type];
		}
	}, {
		key: "getComponents",


		/**
   * getComponents
   * @returns {Object}
   */
		value: function getComponents() {
			return this._components;
		}
	}, {
		key: "setEnabled",


		/**
  * setEnabled
  * @param {bool} enable.
  */
		value: function setEnabled(enable) {
			this._enabled = enable;
		}
	}, {
		key: "isEnabled",


		/**
  * setEnabled
  * @returns {boolean}
  */
		value: function isEnabled() {
			return this._enabled;
		}
	}, {
		key: "setName",


		/**
  * setName
  * @param {String} name.
  */
		value: function setName(name) {
			this._name = name;
		}
	}, {
		key: "getName",


		/**
  * getName
  * @returns {String}
  */
		value: function getName() {
			return this._name;
		}
	}]);

	return Node;
}();

global.Node = Node;
module.exports.Node = Node;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}]},{},[1]);
