// scejs/src/scejs/UI/PanelStage.class.js
/**
* @class
* @constructor
*/
export class PanelStage {
    constructor() {
        this.panel = new StormPanel({"id": 'DIVID_PanelStage',
                                    "paneltitle": 'STAGE'});

        this.elContent = $('#DIVID_PanelStage_content');
        this.elContentDiv = $('#DIVID_PanelStage_content div');
    }

	/**
	 * @callback PanelStage~onselect
	 * @param {Node} node
	 */
	/**
	* show
	* @param {Array<Node>} nodes
	* @param {Node} selectedNode
	* @param {PanelStage~onselect} onselect
	*/
	show(nodes, selectedNode, onselect) {
        this.panel.show();
		
		this.showListObjects(nodes, selectedNode, onselect);
	};
	
	/**
	* showListObjects
	* @param {Array<Node>} nodes
	* @param {Node} selectedNode
	* @param {PanelStage~onselect} onselect
	* @private
	*/
	showListObjects(nodes, selectedNode, onselect) {
        this.elContent.html("");
		let str = '';
		for(let n=0, f = nodes.length; n < f; n++) {
            let colorBg = (selectedNode !== undefined && selectedNode === nodes[n]) ? '#444' : '#000';
            let colorText = (nodes[n].isEnabled() === true) ? '#FFF': '#999';
            str = "<div id='TDID_StormObjectNum_nodes"+n+"' style='background-color:"+colorBg+";color:"+colorText+";'>"+nodes[n].getName()+"</div>";
            this.elContent.append(str);

            let e = document.getElementById("TDID_StormObjectNum_nodes"+n);
            e.addEventListener("click", (function(e, n) {
                onselect(nodes[n]);

                select($(e));
            }).bind(this, e, n));
		}

        this.elContentDiv.css({	'cursor':'pointer',
                            'border':'1px solid #444'});

        this.elContentDiv.bind('mouseover', function() {
            $(this).css({'border':'1px solid #CCC'});
        });

        this.elContentDiv.bind('mouseout', function() {
            $(this).css({'border':'1px solid #444'});
        });
	};

	/**
	* select
	* @param {HTMLDivElement} element
	* @private
	*/
	static select(element) {
        this.elContentDiv.css("background-color","#000");

		if(element !== undefined)
		    element.css("background-color","#444");
	};
}
global.PanelStage = PanelStage;
module.exports.PanelStage = PanelStage;
