// scejs/src/scejs/UI/UI.class.js
import {PanelStage} from "./PanelStage.class";
import {PanelNode} from "./PanelNode.class";
import {PanelNumberGenerator} from "./PanelNumberGenerator.class";
/**
* @class
* @constructor
*/
export class UI {
    constructor(project) {
        this._project = project;
        this.panel_Stage = null;
        this.panel_Node = null;
        this.panel_NumberGenerator = null;
    }

	/**
	 * render
	 * @param {HTMLDivElement} target
     * @param {HTMLCanvasElement} canvas
	 */
	render(target, canvas) {
		let str = ''+
		'<ul id="TopMenu_'+target.id+'">'+
	    	'<li>'+
	    		'Data'+
		        '<ul>'+
		            '<li id="TOPMENU_'+target.id+'_openFile">Open File...</li>'+
		        '</ul>'+
		    '</li>'+
		    '<li>'+
		        'View..'+
		        '<ul>'+
			        '<li class="sub">'+
			            'Panels'+
			            '<ul>'+
			                '<li id="TOPMENU_'+target.id+'_view_PanelStage">Panel Stage...</li>'+
			                '<li id="TOPMENU_'+target.id+'_view_PanelNode">Panel Node...</li>'+
			                '<li id="TOPMENU_'+target.id+'_view_PanelNumberGenerator">Panel Number Generator...</li>'+
			            '</ul>'+
			        '</li>'+
		        '</ul>'+
		    '</li>'+
		'</ul>';
		//target.innerHTML = "";
		new ActionHelpers().appendStringChild(str, target, "init");
		
		$('#TopMenu_'+target.id).fileMenu({
	        slideSpeed: 0
	    });
		//document.getElementById("TopMenu_"+target.id).style.position = "absolute";
        document.getElementById("TopMenu_"+target.id).style.width = canvas.clientWidth+"px";
		document.getElementById("TopMenu_"+target.id).style.fontSize = "8px";
		document.getElementById("TopMenu_"+target.id).style.padding = "2px";
		
		//fileLoader.render(document.body);

        this.panel_Stage = new PanelStage();
        this.panel_Node = new PanelNode();
        this.panel_NumberGenerator = new PanelNumberGenerator();
		
		
		// controllers
        target.querySelector("#TopMenu_"+target.id).addEventListener("mouseover", () => {
			//this._project.getActiveStage().pause();
		});
        target.querySelector("#TopMenu_"+target.id).addEventListener("mouseout", () => {
			//this._project.getActiveStage().render();
		});


        target.querySelector("#TOPMENU_"+target.id+"_openFile").addEventListener("click", () => {
			//fileLoader.openDialog(); 
		});

        target.querySelector("#TOPMENU_"+target.id+"_view_PanelStage").addEventListener("click", (e) => {
			e.preventDefault();
			e.stopPropagation();
            this.panel_Stage.show(this._project.getActiveStage().getNodes(), this._project.getActiveStage().getSelectedNode(), (node) => {
                this._project.getActiveStage().setSelectedNode(node);

                this.panel_Node.show(this._project.getActiveStage().getSelectedNode());
			});
		});

        target.querySelector("#TOPMENU_"+target.id+"_view_PanelNode").addEventListener("click", (e) => {
			e.preventDefault();
			e.stopPropagation();
            this.panel_Node.show(this._project.getActiveStage().getSelectedNode());
		});

        target.querySelector("#TOPMENU_"+target.id+"_view_PanelNumberGenerator").addEventListener("click", (e) => {
			e.preventDefault();
			e.stopPropagation();
            this.panel_NumberGenerator.show();
		});
	}; 
}
global.UI = UI;
module.exports.UI = UI;
