// scejs/src/scejs/ComponentMouseEvents.class.js
import {Component} from "./Component.class";
import {Constants} from "./Constants";

/**
* @class
* @constructor
*/
export class ComponentMouseEvents extends Component {
    constructor() {
        super();
	
        this.type = Constants.COMPONENT_TYPES.MOUSE_EVENTS;
        this.node = null;
        this.gl = null;


        this._onmousedown = null;
        this._onmouseup = null;
        this._onmousemove = null;
        this._onmousewheel = null;
    }
	
	/**
	 * initialize
	 * @param {Node} nod
	 * @param {WebGLRenderingContext} glCtx.
	 * @override
	 */
	initialize(nod, glCtx) {
        this.node = nod;
        this.gl = glCtx;
	};	
	
	/**
	 * @callback ComponentMouseEvents~onmousedown
	 * @param {MouseEvent} evt
	 */	
	/**
	 * onmousedown
	 * @param {ComponentMouseEvents~onmousedown} cb
	 */
	onmousedown(cb) {
		this._onmousedown = cb;
	};	
	
	/**
	 * @callback ComponentMouseEvents~onmouseup
	 * @param {MouseEvent} evt
	 */	
	/**
	 * onmousedown
	 * @param {ComponentMouseEvents~onmouseup} cb
	 */
	onmouseup(cb) {
		this._onmouseup = cb;
	};	
	
	/**
	 * @callback ComponentMouseEvents~onmousemove
	 * @param {MouseEvent} evt
	 * @param {StormV3} dir
	 */	
	/**
	 * onmousedown
	 * @param {ComponentMouseEvents~onmousemove} cb
	 */
	onmousemove(cb) {
		this._onmousemove = cb;
	};	
	
	/**
	 * @callback ComponentMouseEvents~onmousewheel
	 * @param {MouseWheelEvent} evt
	 * @param {StormV3} dir
	 */	
	/**
	 * onmousedown
	 * @param {ComponentMouseEvents~onmousewheel} cb
	 */
	onmousewheel(cb) {
		this._onmousewheel = cb;
	};	
}
global.ComponentMouseEvents = ComponentMouseEvents;
module.exports.ComponentMouseEvents = ComponentMouseEvents;
