// scejs/src/scejs/Component_GPU.class.js
import "webclgl";
import {Component} from "./Component.class";
import {Constants} from "./Constants";

/**
 * @class
 * @constructor
 */
export class Component_GPU extends Component {
    constructor() {
        super();

        this.type = Constants.COMPONENT_TYPES.GPU;
        this.node = null;
        this.gl = null;

        this.gpufG = null;
        this.args = {};
    }


    /**
     * initialize
     * @param {Node} nod
     * @param {WebGLRenderingContext} glCtx.
     * @override
     */
    initialize(nod, glCtx) {
        this.node = nod;
        this.gl = glCtx;
    };

    /**
     * tick
     * @override
     */
    tick(delta) {
        this.tickArguments();

        if(this.gpufG != null)
            this.gpufG.processKernels();

        if(this.gpufG != null)
            this.gpufG.processGraphic(undefined);
    };

    /**
     * setGPUFor
     */
    setGPUFor() {
        for(let key in arguments[1]) {
            let expl = key.split(" ");
            if(expl != null && expl.length > 1) {
                let argName = expl[1];
                this.args[argName] = {
                	"fnvalue": arguments[1][key],
                    "updatable": null,
                    "splits": null,
                    "overrideDimensions": null};
            }
            arguments[1][key] = arguments[1][key]();
        }

        let F = () => {
            return gpufor.apply(this, arguments);
        };
        this.gpufG = new F();
    };


    /**
     * getWebCLGL
     * @returns {WebCLGL}
     */
    getWebCLGL() {
        return this.gpufG.getWebCLGL();
    };


    // █████╗ ██████╗  ██████╗ ██╗   ██╗███╗   ███╗███████╗███╗   ██╗████████╗███████╗
    //██╔══██╗██╔══██╗██╔════╝ ██║   ██║████╗ ████║██╔════╝████╗  ██║╚══██╔══╝██╔════╝
    //███████║██████╔╝██║  ███╗██║   ██║██╔████╔██║█████╗  ██╔██╗ ██║   ██║   ███████╗
    //██╔══██║██╔══██╗██║   ██║██║   ██║██║╚██╔╝██║██╔══╝  ██║╚██╗██║   ██║   ╚════██║
    //██║  ██║██║  ██║╚██████╔╝╚██████╔╝██║ ╚═╝ ██║███████╗██║ ╚████║   ██║   ███████║
    //╚═╝  ╚═╝╚═╝  ╚═╝ ╚═════╝  ╚═════╝ ╚═╝     ╚═╝╚══════╝╚═╝  ╚═══╝   ╚═╝   ╚══════╝

    /**
     *
     */
    addArgument(arg, value) {
        this.args[arg.split(" ")[1]] = {
            "fnvalue": value,
            "updatable": null,
            "splits": null,
            "overrideDimensions": null};

        this.gpufG.addArg(arg);
    };

    /**
     * @param {String} argument Argument to set
     * @param {Function} fnvalue
     * @param {Array<number>} [splits=[array.length]]
     * @param {Array<number>} [overrideDimensions=new Array(){Math.sqrt(value.length), Math.sqrt(value.length)}]
     * @returns {WebCLGLBuffer}
     */
    setArg(argument, fnvalue, splits, overrideDimensions) {
        let buff = this.gpufG.setArg(argument, fnvalue(), splits, overrideDimensions);
        this.args[argument] = {	"fnvalue": fnvalue,
            "updatable": null,
            "splits": splits,
            "overrideDimensions": overrideDimensions};

        return buff;
    };

    /**
     * getComponentBufferArg
     * @param {String} argument Argument to set
     * @param {Component|Component_GPU} comp_gpu
     */
    getComponentBufferArg(argument, comp_gpu) {
        this.gpufG.getGPUForArg(argument, comp_gpu.gpufG);
        this.args[argument] = {	"fnvalue": null,
            "updatable": null,
            "splits": null,
            "overrideDimensions": null};
    };

    /**
     * getArgs
     * @returns {Object}
     */
    getArgs() {
        return this.args;
    };

    /**
     * getAllArgs
     * @returns {Object}
     */
    getAllArgs() {
        return this.gpufG.getAllArgs();
    };

    /**
     * getBuffers
     * @returns {{}|Array<WebCLGLBuffer>}
     */
    getBuffers() {
        return this.gpufG._argsValues;
    };

    /**
     * @param {String} argument Argument to set
     * @param {boolean} value
     */
    setArgUpdatable(argument, value) {
        this.args[argument].updatable = value;
    };

    /**
     * tickArguments
     */
    tickArguments() {
        for(let key in this.args) {
            if(this.args[key].updatable === true) {
                let arg = this.args[key];
                this.gpufG.setArg(key, arg.fnvalue(), arg.splits, arg.overrideDimensions);
            }
        }
    };

}
global.Component_GPU = Component_GPU;
module.exports.Component_GPU = Component_GPU;
