// scejs/src/scejs/ComponentKeyboardEvents.class.js
import {Component} from "./Component.class";
import {Constants} from "./Constants";

/**
* @class
* @constructor
*/
export class ComponentKeyboardEvents extends Component {
    constructor() {
        super();
	
        this.type = Constants.COMPONENT_TYPES.KEYBOARD_EVENTS;
        this.node = null;
        this.gl = null;

        this._onkeydown = null;
        this._onkeyup = null;
    }

	/**
	 * initialize
	 * @param {Node} nod
	 * @param {WebGLRenderingContext} glCtx.
	 * @override
	 */
	initialize(nod, glCtx) {
        this.node = nod;
        this.gl = glCtx;
	};	
	
	/**
	 * @callback ComponentKeyboardEvents~onkeydown
	 * @param {KeyboardEvent} evt
	 */	
	/**
	 * onkeydown
	 * @param {ComponentKeyboardEvents~onkeydown} cb
	 */
	onkeydown(cb) {
		this._onkeydown = cb;
	};		
	
	/**
	 * @callback ComponentKeyboardEvents~onkeyup
	 * @param {KeyboardEvent} evt
	 */	
	/**
	 * onkeyup
	 * @param {ComponentKeyboardEvents~onkeyup} cb
	 */
	onkeyup(cb) {
		this._onkeyup = cb;
	};	
	
}
global.ComponentKeyboardEvents = ComponentKeyboardEvents;
module.exports.ComponentKeyboardEvents = ComponentKeyboardEvents;
