// scejs/src/scejs/Node.class.js
/**
* @class
* @constructor
*/
export class Node {
    constructor() {
        this._components = {};
        this._name = null;
        this._gl = null;

        this._enabled = true;

        this.onTick = null;
    }

    /**
     * initialize
     * @param {String} name.
     * @param {WebGLRenderingContext} glCtx.
     */
    initialize(name, glCtx) {
        this._name = name;
        this._gl = glCtx;
    };

	/**
	 * addComponent
	 * @param {Component} component
	 */
	addComponent(component) {
        this._components[component.type] = component;
		this._components[component.type].initialize(this, this._gl);
	};
	
	/**
	 * getComponent
	 * @param {Constants.COMPONENT_TYPES|int} type
	 * @returns {Component}
	 */
	getComponent(type) {
		return this._components[type];
	};
	
	/**
	 * getComponents
	 * @returns {Object}
	 */
	getComponents() {
		return this._components;
	};
	
	/**
	* setEnabled
	* @param {bool} enable.
	*/
	setEnabled(enable) {
        this._enabled = enable;
	};
	
	/**
	* setEnabled
	* @returns {boolean}
	*/
	isEnabled() {
		return this._enabled;
	};
	
	/**
	* setName
	* @param {String} name.
	*/
	setName(name) {
        this._name = name;
	};
	
	/**
	* getName
	* @returns {String}
	*/
	getName() {
        return this._name;
	};
}
global.Node = Node;
module.exports.Node = Node;
