// scejs/src/scejs/SCE.class.js
/*
The MIT License (MIT)

Copyright (c) <2016> <Roberto Gonzalez. http://stormcolour.appspot.com/>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 */

import {SystemEvents} from "./SystemEvents.class";
import {Utils} from "./Utils.class";
import {Constants} from "./Constants";



/*var includesF = [
				'/UI/UI.class.js',
				'/UI/PanelStage.class.js',
				'/UI/PanelNode.class.js',
				'/UI/PanelNumberGenerator.class.js',
				'/UI/UIComponentGPU.class.js',
				'/UI/UIComponentGPU_Indices.class.js',
				'/UI/UIComponentGPU_Argument.class.js',
				'/UI/UIComponentGPU_Kernel.class.js',
				'/UI/UIComponentGPU_Vfp.class.js'];*/

//***********
//UI LIBRARIES
//***********
//CSS
/*if(window.jQuery == undefined) {
	document.write('<link rel="stylesheet" type="text/css" href="'+sceDirectory+'/UI/JQuery/ui/jquery-ui.min.css" />');
}
document.write('<link rel="stylesheet" type="text/css" href="'+sceDirectory+'/UI/style.css" />');
document.write('<link rel="stylesheet" type="text/css" href="'+sceDirectory+'/UI/fileMenu/fileMenu.css" />');
document.write('<link rel="stylesheet" type="text/css" href="'+sceDirectory+'/UI/stormPanel/stormPanel.css" />');

//JS
if(window.jQuery == undefined) {
	document.write('<script type="text/javascript" src="'+sceDirectory+'/UI/JQuery/jquery-1.11.3.js"></script>');	
	document.write('<script type="text/javascript" src="'+sceDirectory+'/UI/JQuery/ui/jquery-ui.min.js"></script>');
}
document.write('<script type="text/javascript" src="'+sceDirectory+'/UI/fileMenu/fileMenu.js"></script>');
document.write('<script type="text/javascript" src="'+sceDirectory+'/UI/stormPanel/stormPanel.js"></script>');
document.write('<script type="text/javascript" src="'+sceDirectory+'/UI/ActionHelpers.class.js"></script>');*/

/**
* Engine contructor
* @class
* @constructor
*/
export class SCE {
    constructor() {
        this.target = null;
        this.project = null;
        this.dimensions = null;
        this.canvas = null;
        this.gl = null;
        this._UI = null;
        this._enableUI = null;
        this._events = null;

        this.sceDirectory = document.querySelector('script[src$="SCE.class.js"]');
        if(this.sceDirectory === null)
            this.sceDirectory = document.querySelector('script[src$="SCEJS.min.js"]');
        this.sceDirectory = this.sceDirectory.getAttribute('src');

        let page = this.sceDirectory.split('/').pop();
        this.sceDirectory = this.sceDirectory.replace('/'+page,"");
    }

	/**
	* @param {Object} jsonIn
	* @param {HTMLDivElement} jsonIn.target
	* @param {Object} [jsonIn.dimensions={width: Int, height: Int}]
	* @param {boolean} [jsonIn.enableUI=false]
    * @param {WebGLRenderingContext} [jsonIn.gl=undefined]
	*/
	initialize(jsonIn) {
        this.target = (jsonIn.target !== undefined) ? jsonIn.target : undefined;
        this.dimensions = (jsonIn.dimensions !== undefined) ? jsonIn.dimensions : {"width": 512, "height": 512};
        this._enableUI = (jsonIn.enableUI !== undefined) ? jsonIn.enableUI : false;
			
		if(this.target != null) {
            this.canvas = document.createElement("canvas");
            this.target.appendChild(this.canvas);
			this.setDimensions(this.dimensions.width, this.dimensions.height);

			if(jsonIn.gl !== undefined && jsonIn.gl !== null) {
                this.gl = jsonIn.gl;
            } else if(!(this.gl = Utils.getWebGLContextFromCanvas(this.canvas))) {
				alert("No WebGLRenderingContext");
				return false; 
			}
		} else alert('Target DIV required');
	};
	
	/**
	 * loadProject 
	 * @param {Project} prj
	 */
	loadProject(prj) {
        this.project = prj;
        this.project.setWebGLContext(this.gl);
		
		if(this._enableUI === true)
            this._UI = new UI(this.project).render(this.target, this.canvas);

        this._events = new SystemEvents(this, this.canvas);
        this._events.initialize();
	};
	
	/**
	 * getLoadedProject 
	 * @returns {Project}
	 */
	getLoadedProject() {
		return this.project;
	};
	
	/**
	 * getCanvas
	 * @returns {HTMLCanvasElement}
	 */
	getCanvas() {
		return this.canvas;
	};
	
	/**
	 * getEvents
	 * @returns {SystemEvents}
	 */
	getEvents() {
		return this._events;
	};
	
	/**
	 * setDimensions
	 * @param {int} width
	 * @param {int} height
	 */
	setDimensions(width, height) {
        this.dimensions = {"width": width, "height": height};

        this.canvas.setAttribute("width", this.dimensions.width);
        this.canvas.setAttribute("height", this.dimensions.height);


		if(this.project != null) {
		    if(this.project.getActiveStage() != null) {
                let actCamera = this.project.getActiveStage().getActiveCamera();
		        if(actCamera != null) {
                    let comp_proj = actCamera.getComponent(Constants.COMPONENT_TYPES.PROJECTION);
                    let comp_screenEffects = actCamera.getComponent(Constants.COMPONENT_TYPES.GPU);

                    comp_proj.setResolution(this.dimensions.width, this.dimensions.width);
                    comp_screenEffects.setArg("RGB", (function() {return new Float32Array(this.dimensions.width*this.dimensions.width*4);}).bind(this));
                }
            }
		}
	};
	
	/**
	 * @typedef {Object} SCE~getDimensions
	 * @property {int} SCE~getDimensions.width
	 * @property {int} SCE~getDimensions.height
	 */
	/**
	 * getDimensions
	 * @returns {SCE~getDimensions}
	 */
	getDimensions() {
		return this.dimensions;
	};
}
global.SCE = SCE;
module.exports.SCE = SCE;
